lager/lenses/tuple.hpp

```diff
@@ -0,0 +1,194 @@
+#pragma once
+
+#include <tuple>
+#include <type_traits>
+
+#include <lager/lenses.hpp>
+#include <lager/lenses/attr.hpp>
+
+namespace lager {
+namespace lenses {
+namespace detail {
+
+template <typename T>
+struct templated_type_contructor;
+
+template <template <typename...> typename T, typename... Params>
+struct templated_type_contructor<T<Params...>>
+{
+    template <typename... Ts>
+    static auto construct(Ts&&... ps)
+    {
+        return T<std::remove_reference_t<Ts>...>{std::forward<Ts>(ps)...};
+    }
+};
+
+template <size_t I, typename... Tuples>
+decltype(auto) multiget(Tuples&&... tuples)
+{
+    return std::forward_as_tuple(std::get<I>(std::forward<Tuples>(tuples))...);
+};
+
+template <typename ResType, typename F, typename... Tuples, size_t... Seq>
+decltype(auto) apply_zip_impl(std::index_sequence<Seq...>, F&& f, Tuples&&... t)
+{
+    return templated_type_contructor<ResType>::construct(std::apply(
+        std::forward<F>(f), multiget<Seq>(std::forward<Tuples>(t)...))...);
+}
+
+template <typename ResType, typename F, typename... Tuple>
+decltype(auto) apply_zip(F&& f, Tuple&&... t)
+{
+    constexpr size_t len =
+        std::min({std::tuple_size_v<std::remove_reference_t<Tuple>>...});
+    return apply_zip_impl<ResType>(std::make_index_sequence<len>{},
+                                   std::forward<F>(f),
+                                   std::forward<Tuple>(t)...);
+}
+
+namespace fold_op {
+template <typename Init, typename Elem>
+decltype(auto) operator+(Init&& init, Elem&& elem)
+{
+    auto&& [lens, part] = LAGER_FWD(elem);
+    return ::lager::set(LAGER_FWD(lens), LAGER_FWD(init), LAGER_FWD(part));
+}
+} // namespace fold_op
+
+template <typename Whole, typename LensTuple, typename PartTuple, size_t... Seq>
+decltype(auto) set_fold_impl(Whole&& whole,
+                             LensTuple&& lenses,
+                             PartTuple&& parts,
+                             std::index_sequence<Seq...>)
+{
+    using fold_op::operator+;
+    return (LAGER_FWD(whole) + ... +
```

Comment:
This use of folding expressions is very smart, it had me smiling at the screen :)

However, it might be simpler to use recursion, so you don't have to define the `operator+` and `multiget` separatelly. (But maybe it's fine like this, I dunno.)

Reply:
I just prefer fold expressions over recursion for this kind of thing, it should improve compile times at least marginally.

Reply:
Sounds good.